blackbird_python/blackbird/tests/test_listener.py

```diff
@@ -346,6 +347,313 @@ def test_target_arg(self, parse_input):
+            {"op": "Rgate", "args": [0.1], "kwargs": {}, "modes": [2]},
+        ]
+
+        assert bb.operations == expected
+
+    @pytest.mark.skipif(sys.version_info < (3, 6), reason="tmpdir fixture requires Python >=3.6")
+    def test_nested_include(self, parse_input, tmpdir):
+        """Test nested includes"""
+        template1 = textwrap.dedent(
+            """
+            name MachZehnder
+            version 0.0
+            BSgate({theta}, pi/2) | [0, 1]
+            Rgate({phi}) | 0
+            """
+        )
+
+        filename1 = tmpdir.join("mach_zehnder.xbb")
+
+        with open(filename1, "w") as f:
+            f.write(template1)
+
+        template2 = textwrap.dedent(
+            """
+            name Chip15
+            version 0.0
+            include "{}"
+            S2gate({{sq}}) | [0, 1]
+            MachZehnder(theta={{theta}}, phi={{phi}}) | [1, 2]
+            """
+        ).format(filename1)
+
+        filename2 = tmpdir.join("chip16.xbb")
+
+        with open(filename2, "w") as f:
+            f.write(template2)
+
+        test_include = textwrap.dedent(
+            """
+            name test_include
+            version 0.0
+            include "{}"
+            Chip15(sq=0.5, theta=0.54, phi=0.1) | [0, 2, 1]
+            """
+        ).format(filename2)
+
+        bb = parse_input(test_include, cwd=tmpdir)
+
+        expected = [
+            {"op": "S2gate", "args": [0.5], "kwargs": {}, "modes": [0, 2]},
+            {"op": "BSgate", "args": [0.54, np.pi / 2], "kwargs": {}, "modes": [2, 1]},
+            {"op": "Rgate", "args": [0.1], "kwargs": {}, "modes": [2]},
+        ]
+
+        assert bb.operations == expected
+
+    @pytest.mark.skipif(sys.version_info < (3, 6), reason="tmpdir fixture requires Python >=3.6")
+    def test_multiple_nested_include(self, parse_input, tmpdir):
+        """Same as the above, but tests the case where the same file
+        might be imported multiple times in different places."""
+        template1 = textwrap.dedent(
+            """
+            name MachZehnder
+            version 0.0
+            BSgate({theta}, pi/2) | [0, 1]
+            Rgate({phi}) | 0
+            """
+        )
+
+        filename1 = tmpdir.join("mach_zehnder.xbb")
+
+        with open(filename1, "w") as f:
+            f.write(template1)
+
+        template2 = textwrap.dedent(
+            """
+            name Chip15
+            version 0.0
+            include "{}"
+            S2gate({{sq}}) | [0, 1]
+            MachZehnder(theta={{theta}}, phi={{phi}}) | [1, 2]
+            """
+        ).format(filename1)
+
+        filename2 = tmpdir.join("chip16.xbb")
+
+        with open(filename2, "w") as f:
+            f.write(template2)
+
+        test_include = textwrap.dedent(
+            """
+            name test_include
+            version 0.0
+            include "{}"
+            include "{}"
+            Chip15(sq=0.5, theta=0.54, phi=0.1) | [0, 2, 1]
+            """
```

Comment:
Everything is ok here, but what if `MachZender` was also used within "test_include"? In particular, what if there was a conflict with two different versions of `MachZender` ending up in the same namespace due to earlier `include`s?

Reply:
> but what if MachZender was also used within "test_include"?

That would also work! 🙂 the parser is smart and only parses included files once even if included in multiple places, but still makes it available everywhere the user has placed the `include`.

> two different versions of MachZender ending up in the same namespace due to earlier includes?

In this case, the second/last include will overwrite the previous one. Similar to Python/C++ if you import/include different modules/headers within the same namespace.